docs/retry_timeout.rst

```diff
@@ -0,0 +1,114 @@
+Configuring Timeouts and Retries
+================================
+
+When using object methods which invoke Google Cloud Storage API methods,
+you have several options for how the library handles timeouts and
+how it retries transient errors.
+
+
+.. _configuring_timeouts:
+
+Configuring Timeouts
+--------------------
+
+For a number of reasons, methods which invoke API methods may take
+longer than expected or desired.  By default, such methods all time out
+after a default interval, 60.0 seconds.  Rather than blocking your application
+code for that interval, you may choose to configure explicit timeouts
+in your code, using one of three forms:
+
+- You can pass a single integer or float which functions as the timeout for the
+  entire request. E.g.:
+
+.. code-block:: python
+
+   bucket = client.get_bucket(BUCKET_NAME, timeout=300.0)  # five minutes
+
+- You can also be passed as a two-tuple, ``(connect_timeout, read_timeout)``,
+  where the ``connect_timeout`` sets the maximum time required to establish
+  the connection to the server, and the ``read_timeout`` sets the maximum
+  time to wait for a completed response.  E.g.:
+
+.. code-block:: python
+
+   bucket = client.get_bucket(BUCKET_NAME, timeout=(3, 10))
+
+
+- You can also pass ``None`` as the timeout value:  in this case, the library
+  will block indefinitely for a response.  E.g.:
+
+.. code-block:: python
+
+   bucket = client.get_bucket(BUCKET_NAME, timeout=None)
+
+.. note::
+   Depending on the retry strategy, a request may be
+   repeated several times using the same timeout each time.
+
+See also:
+
+  :ref:`Timeouts in requests <requests:timeouts>`
+
+
+.. _configuring_retries:
+
+Configuring Retries
+--------------------
+
+Methods which invoke API methods may fail for a number of reasons, some of
+which represent "transient" conditions, and thus can be retried
+automatically.  The library tries to provide a sensible default retry policy
+for each method, base on its semantics.
```

Comment:
I would add a little more color on the default retry policy at the top here. Generally, the strategy is to retry any idempotent operation (some calls are always idempotent while others require a precondition to qualify) on any transient error.

Reply:
4278187